[프레디]미션3: 체스판 초기화

Dae-Hwa opened the pull request:
안녕하세요! 프레디입니다.
리모트 브랜치에 푸쉬가 잘못되어 수정 후 다시 요청드립니다.

**Pawn**

- 검은색은 대문자, 흰색은 소문자 p를 반환하는 `getRepresentation()` 을 추가했습니다. 유니코드 체스 기호가 있어 사용을 고려해봤는데 터미널에서는 검은색과 흰색의 분간이 힘들어 요구사항에 적힌대로 구현했습니다.

**Board**

- 초기화 할 수 있는 `initialize()` 를 추가했습니다. 7번 rank(files의 1번 row)에 검은색, 2번 rank(files의 6번 row)에 흰색 폰을 배치시켰습니다.
- 초기화된 흰색 폰과 검은색 폰을 호출할 수 있는 `getWhitePawnsResult()`와 `getBlackPawnsResult()`을 추가했습니다.
- 프린트 기능을 추가했습니다. 빈 줄은 `........`, 폰이 있는 줄은 폰을 출력합니다.

**File**

- Board에서 초기화 할 때 사용되는 `fillWith(Color)` 를 추가했습니다. 인자로 받은 색깔로 해당 file을 채워줍니다.
- 선택한 file을 프린트 하는데 비어있는 file이면 `........`을 출력하는 `getRepresetation()`을 추가했습니다.

**Main**

- 입출력을 받아 체스게임을 실행시키는 `Main`을 추가했습니다.
- 입력을 받아 start일 경우 초기화, end일 경우 종료합니다.

감사합니다!

Dae-Hwa committed:
Add getRepresentation at Pawn

검은색은 대문자, 흰색은 소문자 p 반환

Dae-Hwa committed:
Fix remove unnecessary public modifier in BoardTest

Dae-Hwa committed:
Add initialize that use getPawnsResult in Board

초기화시 a7~h7에는 검은 폰, a2~h2에는 흰색 폰을 추가
getWhitePawnsResult와 getBlackPawnsResult를 이용하여 불러올 수 있음

Dae-Hwa committed:
Refactor initialize and getPawnsResult in Board

getBlackFile과 getWhiteFile 추가하여 중복 제거

Dae-Hwa committed:
Add fillWith in File

Dae-Hwa committed:
Refactor initialize in Board to use fillWith of File

Dae-Hwa committed:
Add getRepresentation in File

Dae-Hwa committed:
Add print in Board

Dae-Hwa committed:
Create Main

체스게임을 실행시키는 메인 클래스
입력을 받아 start일 경우 초기화하고 출력
end일 경우 게임 종료

Dae-Hwa committed:
Refactor to use assertAll if we can replace

Dae-Hwa committed:
Fix a lint error

Dae-Hwa committed:
Refactor getRepresentation in File to use isEmpty

size == 0 으로 판단하는 것 isEmpty로 변경

Dae-Hwa committed:
Refactor getRepresentation in File to use stream

for-each와 StringBuilder로 문자열 합치던 것 스트림으로 변경

tmdgusya commented on src/main/java/kr/codesquad/freddie/chess/Main.java:
```diff
@@ -0,0 +1,23 @@
+
+public class Main {
+    public static void main(String[] args) throws IOException {
+        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
```
개인적으로는 try-with-resources 를 사용해서 br 객체의 사용이 끝났을때 
close() 가 되도록 해줬으면 어땠을까 라는 생각이 드네요 :)

생각해보니 저도 적지않은것 같네요 ㅋㅋ 프레디 코드 리뷰해주다가 기억났습니다!

ksundong replied:
좋은 리뷰네요. 혹시 `BufferedReader`를 따로 사용해주신 이유가 있나요?

ksundong replied:
왜 `close()`를 해줘야하는지도 학습해보시구요. `BufferdReader`만 `close()`해줘도 될까요?
된다면 그 이유도 한 번 알아보세요!

Dae-Hwa replied:
우선 `BufferedReader`는 일반적으로 `Scanner`보다 속도가 빠르다고 알고 있습니다. 규모가 크지 않아 큰 영향은 없겠지만 궁금해서 찾아봤는데, `BufferedReader`는 일반적으로 IO를 다루듯이 바이트 단위로 가져오는데, `Scanner`는 버퍼 크기만큼 잘라서 정규표현식을 실행하는 것을 확인했습니다. 속도에 민감하다면 `BufferedReader`를 사용하는게 훨씬 유리할 것 같습니다.

둘째로 다루기가 쉽다고 생각해서 사용했습니다. `Scanner`의 경우 문자열을 토큰단위로 쉽게 사용할 수 있게 만들어놨지만, 커맨드 라인 프로그램을 만들 때 대부분의 입력이 줄 단위로 들어오기 때문에 결국 line단위로 끊어서 문자열을 새로 컨트롤 하는것이 더 편하게 느껴졌습니다. 그럴 경우 위에서 말씀드린 이유로 `BufferedReader`를 쓰는게 더 낫다는 생각으로 사용했습니다.

Dae-Hwa replied:
`BufferedReader`의 `close()` 가 내부 필드인 `InputStreamReader` 의 `close()` 를 실행시켜주고 마찬가지로 계속해서 상위 클래스의 `close()` 를 호출해주는 것을 확인했습니다. 따라서 `BufferedReader`만 닫아줘도 무방할 것 같습니다.

`close()`는 main이 종료되면 JVM도 종료될테니 자동으로 리소스 할당이 없어질거라 생각하여 하지 않았었는데, 대부분의 경우나 오라클의 예제에서도 close를 사용하는걸 보아 무조건 하는게 좋다!라고 결론을 내렸습니다.

JVM이 종료되어도 리소스 할당이 해제되지 않는 경우를 생각해봤는데 IDE는 닫혔는데 JVM 실행은 정상적으로 되지 않았을 경우 고아 프로세스가 발생할 수 있겠다는 생각을 해봤습니다. 혹시 이외에도 예시가 떠오르시면 공유 부탁드립니다!🙏
운영체제의 어떤 리소스가 할당되는지도 알아보려고 했는데 여기까지는 이해가 잘 안되네요 😞

Dae-Hwa replied:
Roach가 말해준대로 try-with-resource 사용하여 개선했습니다.
cb7a6c1

ksundong replied:
> 우선 `BufferedReader`는 일반적으로 `Scanner`보다 속도가 빠르다고 알고 있습니다. 규모가 크지 않아 큰 영향은 없겠지만 궁금해서 찾아봤는데, `BufferedReader`는 일반적으로 IO를 다루듯이 바이트 단위로 가져오는데, `Scanner`는 버퍼 크기만큼 잘라서 정규표현식을 실행하는 것을 확인했습니다. 속도에 민감하다면 `BufferedReader`를 사용하는게 훨씬 유리할 것 같습니다.
> 
> 둘째로 다루기가 쉽다고 생각해서 사용했습니다. `Scanner`의 경우 문자열을 토큰단위로 쉽게 사용할 수 있게 만들어놨지만, 커맨드 라인 프로그램을 만들 때 대부분의 입력이 줄 단위로 들어오기 때문에 결국 line단위로 끊어서 문자열을 새로 컨트롤 하는것이 더 편하게 느껴졌습니다. 그럴 경우 위에서 말씀드린 이유로 `BufferedReader`를 쓰는게 더 낫다는 생각으로 사용했습니다.

첫번째는 이론상 맞는 얘기입니다. 문제는 우리가 작성하는 프로그램은 속도와 전혀 무관한 프로그램이죠?
두번째는 취향차이긴 한데, 커맨드라인 프로그램은 라인단위로 동작하기 보다는, 구분자 단위로 동작하는 것이 일반적이지 않나요?
아무튼 고민하시느라 고생하셨습니다. 보통 알고리즘 문제풀이의 관성으로 `BufferedReader`를 많이 사용하는데, 이를 조금 생각하셨으면 좋겠어서 여쭤봤습니다.

ksundong replied:
> `BufferedReader`의 `close()` 가 내부 필드인 `InputStreamReader` 의 `close()` 를 실행시켜주고 마찬가지로 계속해서 상위 클래스의 `close()` 를 호출해주는 것을 확인했습니다. 따라서 `BufferedReader`만 닫아줘도 무방할 것 같습니다.
> 
> `close()`는 main이 종료되면 JVM도 종료될테니 자동으로 리소스 할당이 없어질거라 생각하여 하지 않았었는데, 대부분의 경우나 오라클의 예제에서도 close를 사용하는걸 보아 무조건 하는게 좋다!라고 결론을 내렸습니다.
> 
> JVM이 종료되어도 리소스 할당이 해제되지 않는 경우를 생각해봤는데 IDE는 닫혔는데 JVM 실행은 정상적으로 되지 않았을 경우 고아 프로세스가 발생할 수 있겠다는 생각을 해봤습니다. 혹시 이외에도 예시가 떠오르시면 공유 부탁드립니다!🙏
> 운영체제의 어떤 리소스가 할당되는지도 알아보려고 했는데 여기까지는 이해가 잘 안되네요 😞

네 맞습니다. 잘 찾아보셨네요. 👍 
오라클 공식 튜토리얼을 보면 누수가 발생할 수 있다고 합니다. 저도 누수되는 상황을 겪어보진 않았는데, 일반적으로 자원의 할당은 무조건 닫아주는 것이 좋은 습관입니다.
이를 발생시키는 것도 하나의 좋은 학습이 될 것 같네요.
마지막 내용도 한 번 공부해보면 좋을 것 같습니다!

ksundong replied:
일단 `System.in`이 할당되니까 표준 입력이 할당된다고 보시면 될 것 같아요~

Dae-Hwa replied:
감사합니다! 조금 막막했는데, 말씀해주신 키워드로 찾아봐야겠어요!

tmdgusya commented on src/main/java/kr/codesquad/freddie/chess/board/Board.java:
```diff
@@ -97,4 +98,12 @@ private File getWhiteFile() {
+    public void print() {
+        String result = files.stream()
+                .map(File::getRepresentation)
+                .collect(Collectors.joining(System.lineSeparator()));
```
stream 활용 보기 좋네요.
하나 배워갑니다!

ksundong replied:
Stream을 열심히 활용하시네요 👍

ksundong commented on src/main/java/kr/codesquad/freddie/chess/Main.java:
```diff
@@ -0,0 +1,23 @@
+        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+        Board board = new Board();
+
+        for (String command = br.readLine();
+             !command.equals("end");
+             command = br.readLine()) {
```
와 이런 스타일로 for loop를 쓰시다니... 신기하네요. 👍

Dae-Hwa replied:
혹시 이것도 가독성에는 좋지 않은 패턴일까요?

ksundong replied:
저는 보통 입출력은 while을 썼는데, 이해가 안가는 코드는 아니어서 따봉 드렸습니다 ㅋㅋ

Dae-Hwa replied:
앗..ㅋㅋㅋㅋ 알고리즘을 자바로 풀다보니 이렇게 된 것 같습니다. 😂

kyupid replied:
우와 이거 머에요? 신세계다..🤩 저도 다음에 써보겠습니다 헤헤

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/File.java:
```diff
@@ -41,6 +43,20 @@ public boolean isAddable() {
         return size() < SIZE;
     }
 
+    public void fillWith(Color color) {
+        for (; size() < SIZE; pawns.add(new Pawn(color))) ;
+    }
```
for문의 마법사 같아요.
저라면 for문에는 로직을 넣기 보다는 while을 활용했을 것 같네요.
```suggestion
    public void fillWith(Color color) {
        while (isAddable()) {
            pawns.add(new Pawn(color));
        }
    }
```

너무 축약형만 생각하다보면, 가독성이 떨어지게됩니다.
또한, 중복도 제거해보았어요.

ghojeong replied:
> 저라면 for문에는 로직을 넣기 보다는 while을 활용했을 것 같네요.

동의합니다.
저는 개인적으로 for loop 를 할 때의 조건문에 활용되는 변수는, for loop block 안에 한정되는 지역변수이어야 한다고 생각합니다.

[이펙티브 자바: [아이템 57] 지역변수의 범위를 최소화하라](https://github.com/java-squid/effective-java/tree/master/chapter09/item57)

kyupid replied:
정말 for moon의 마법사이시네요. 전 forEach를 썼어요.

Dae-Hwa replied:
> > 저라면 for문에는 로직을 넣기 보다는 while을 활용했을 것 같네요.
> 
> 동의합니다.
> 저는 개인적으로 for loop 를 할 때의 조건문에 활용되는 변수는, for loop block 안에 한정되는 지역변수이어야 한다고 생각합니다.
> 
> [이펙티브 자바: [아이템 57] 지역변수의 범위를 최소화하라](https://github.com/java-squid/effective-java/tree/master/chapter09/item57)

지역변수는 사용되지 않았다고 생각하는데, 어떤 부분을 말씀해주신건가요?

ksundong replied:
지역변수가 없어서 말씀하신 것 같아요~

ksundong commented on src/main/java/kr/codesquad/freddie/chess/piece/Pawn.java:
```diff
@@ -21,6 +21,10 @@ public Color getColor() {
 
+    public String getRepresentation() {
+        return color == Color.BLACK ? "P" : "p";
+    }
```
enum을 사용하셨으니, enum을 활용해서 representation을 반환하는 방법도 한 번 생각해보세요!

Dae-Hwa replied:
`Color` 에 추상 메소드를 추가하여 각 색상별로 대문자나 소문자를 출력할 수 있도록 하여 연산자를 제거했습니다.
b6ef291
6cbfe52
이럴 경우 커밋을 묶는게 더 좋은가요? 이번에는 커밋 설명에 and가 들어갈 것 같아 나눠봤습니다.

ksundong replied:
저는 묶어주는 편입니다.
커밋은 한 단위의 작업이면 충분하다고 생각해요. 해당 작업을 되돌렸을 때, '한 번에 되돌릴 수 있는가?'도 중요하다고 생각해서요.

Dae-Hwa replied:
판단을 잘 해야겠네요 ㅎㅎ 좋은 의견 감사합니다! 큰 도움이 될 것 같습니다.

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardTest.java:
```diff
@@ -81,4 +81,21 @@ void findPawn_moreThanOne() {
+        board.initialize();
+
+        board.print();
+    }
```
이 테스트는 무엇이 문제일까요?
이 테스트가 성공했는지 실패했는지는 어떻게 판단할까요?

ksundong replied:
보통 출력은 테스트하지 않는 경우가 대부분이고, 출력자체도 assertion이 가능하긴 합니다.

Dae-Hwa replied:
BoardDTO를 새로 생성해서 반환값을 비교하는 테스트를 하도록 수정했습니다.
eae3634

보통 출력은 테스트하지 않는 경우가 대부분이라 하셔서 궁금한 점이 생겼습니다.
지금은 콘솔로 출력을 하니 사용자가 보는 화면이 곧 콘솔에 출력되는 것인데, 이럴 경우 테스트를 하고 싶으면 직접 조작해보며 테스트 하는게 더 좋은 방법인가요?

ksundong replied:
`DTO`는 현재단계에선 다루지 말아봅시다. 커밋만 봐서는 테스트만 추가되어서 DTO가 추가된 배경이 잘 파악이 안되어서 코드 전반적으로 확인해볼게요.
제가 의도한건, `print()`의 책임을 분리해서, `print()`에서 생성된 문자열을 Main에서 출력만하고, `print()`의 문자열 내용을 테스트하는 것을 의도했습니다.

ksundong commented on src/test/java/kr/codesquad/freddie/chess/piece/PawnTest.java:
```diff
@@ -13,4 +14,13 @@ void create() {
+    @DisplayName("검정은 대문자 P 하양은 소문자 p 반환")
+    void getRepresentation() {
+        assertAll(
+                () -> assertThat(Pawn.create("black").getRepresentation()).isEqualTo("P"),
+                () -> assertThat(Pawn.create("white").getRepresentation()).isEqualTo("p")
```
Enum이 굳이 있다면, Enum을 활용해서 생성하는 팩토리 메서드로 만들어주시는게 좋아보여요.
아니면, 각각 팩토리 메서드를 만들어주는 것도 방법일 것 같아요. 일단 편한대로 만들어보세요!
마찬가지로, representation도 Enum을 이용해 변환한다면 보다 깔끔해질 것 같습니다.

Dae-Hwa replied:
enum을 활용해 팩토리 메서드를 만드는 방법을 이렇게 이해했습니다.
```java
public enum Color {
    WHITE {
        @Override
        Pawn createPawn() {
            return new Pawn(WHITE);
        }
    }
...
   abstract Pawn createPawn();
}   
```
Color는 Pawn을 보충해주는 의미로 사용된다고 생각되어서 Pawn을 생성하는 팩토리로 활용하는게 어색하게 느껴지는데, 여기에 관해서 의견을 여쭙고 싶습니다. 그냥 각각의 팩토리 메소드나 새로운 팩토리 객체를 만드는 것을 고려해보다가 혹시 제가 놓치는 부분이 있을까 궁금해졌습니다.

Dae-Hwa replied:
지금은 BalckPawn과 WhitePawn을 생성해주는 팩토리 메소드를 만들어줬습니다.
0e53807

ksundong replied:
앗 이것 역시 제가 의도한건, Pawn에 Enum을 활용한 팩토리 메서드를 만들어주는 것이었습니다.
```
Pawn.create(BLACK).getRepresentation();
```

또는

```
Pawn.createBlack().getRepresentation()
```

같은 느낌으로요.

ksundong replied:
enum에는 말씀하신것처럼 팩토리 메서드를 생성하면 어색해지죠. 제가 잘못 말씀드린 것 같아 죄송하네요 ㅠㅠ

Dae-Hwa replied:
제가 잘못 이해했었네요! ㅎㅎ 답변 감사합니다! enum을 활용하는 것도 고려해봐야겠네요.

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/FileTest.java:
```diff
@@ -84,4 +84,34 @@ void isAddable() {
+
+    private void checkFillWith(Color expected) {
+        for (int i = 0; i < File.SIZE; i++) {
+            assertThat(file.get((char) (i + 'a')).getColor()).isEqualTo(expected);
```
왜 char로 변환해주셨나 했더니, 체스판 자체를 코드에 녹여내신 거군요.
저라면 별도의 메서드로 추출하거나 내부로 옮겨서, 읽는 사람이 좀 더 읽기 편하도록 해줄 것 같아요.

Dae-Hwa replied:
여러가지 고민을 해보다가, 어차피 사용자에게 노출되는 인터페이스는 `Board`이므로 일반적인 인덱스로 접근할 수 있는 `get`을 하나 더 추가해줬습니다.
44803f8

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/Board.java:
```diff
@@ -65,4 +67,43 @@ public String toString() {
+    public void initialize() {
+        getBlackFile().fillWith(Color.BLACK);
+        getWhiteFile().fillWith(Color.WHITE);
+    }
```
딱 해야할 일만 잘 나눠주셨어요.

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/Board.java:
```diff
@@ -65,4 +67,43 @@ public String toString() {
+            sb.append(file.get((char) (i + 'a')).getRepresentation());
+        }
+
+        return sb.toString();
```
테스트 코드랑 어느정도 중복 되는 부분이 보이시나요?
제가 생각했을 때에는 char로 변환되는 부분이 get 안으로 들어가는 편이 더 괜찮아보이는데 어떻게 생각하세요?

Dae-Hwa replied:
만들어둔 다른 메소드를 활용할 수 있도록 수정했습니다.
e22f8a0

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/File.java:
```diff
@@ -1,9 +1,11 @@
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class File {
```
이 클래스의 기본생성자는 직접 만들어주신 이유가 있나요?

Dae-Hwa replied:
뭔가 하려고 했나봐요 😅 제거했습니다!

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/Board.java:
```diff
@@ -65,4 +67,43 @@ public String toString() {
+                .map(File::getRepresentation)
+                .collect(Collectors.joining(System.lineSeparator()));
+
+        System.out.println(result);
```
여기서 print를 해주는 것 보다는 외부에서 출력에 대한 책임을 갖게한다면
결과에 대한 테스트도 가능해지고, 책임이 하나로 명확해지겠죠?

Dae-Hwa replied:
외부와 연결에 사용할 수 있는 `BoardDTO`를 추가로 만들어줬습니다.
하지만 그럼에도 Board를 Main에서 직접 움직이도록 되어있는데요 이는 Main을 컨트롤러처럼 생각했기 때문입니다.
그런데 생각해보니 Main을 클라이언트처럼 생각할 수도 있을 것 같은데 그렇다면 Controller를 사이에 추가해주는 것이 더 좋은 방향일지 생각해보고 있습니다.

ksundong replied:
제 생각엔 굳이 계층을 나눌정도로 프로그램이 복잡하다고 판단되지 않습니다.
따라서 MVC 모델의 방식은 머릿속에서 잠깐 잊어봅시다~ 오히려 코드를 짜는데 방해만 될거에요.

Dae-Hwa replied:
너무 어렵게 생각했나 보네요 😅 '출력에 대한 책임'이라는 말에 집중해서 생각하지 못한 것 같습니다. 답변 감사합니다!

ksundong reviewed (changes requested):
안녕하세요! 프레디!! 리뷰어 디온입니다. 벌써 미션3이라니 정말 대단해요!🎉
본격적으로 미션이 시작되었다는 느낌이 드는 과제였습니다. 고생많으셨어요.
전반적으로 깔끔한 코드를 작성해주셨습니다.
간단한 리뷰내용을 남겨드릴게요. 사실 다음 단계를 바로 진행해도 될 정도로 요구사항을 충실히 지켜주셨어요.
다만 다음단계로 바로 넘어가기 전에 조금만 더 채워봅시다!

## 리뷰

- for문만 사용하기 보다는 while문도 적절히 사용해주시는게 좋아보여요.
- enum을 보다 적극적으로 사용해주세요. enum은 할 수 있는 일이 굉장히 많답니다!
- `File`의 `get`은 읽는 사람 입장에서는 읽기 힘든 코드였어요. 어떻게 하면 읽기 좋게 할 수 있을까요?

고생많으셨습니다! 궁금하신 점은 코멘트 남겨주세요!
수정을 마치시고 리뷰어 다시 지정해주시면 다시 리뷰하겠습니다!

ghojeong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardTest.java:
```diff
@@ -81,4 +81,21 @@ void findPawn_moreThanOne() {
         );
     }
+
+    @Test
+    void initialize() {
```
함수명으로 표현이 되어 있다고 하더라도 DisplayName 을 통해, 무슨 테스트인지 좀더 자세하게 묘사해주시면 좋을 것 같아요

ksundong replied:
이 부분은 저는 딱히 문제 없다고 판단했습니다.

Dae-Hwa committed:
Refactor fillWith in File to use while instead for

for문대신 while 사용하여 가독성 개선. isAddable 이용하여 중복 제거.
https://github.com/codesquad-members-2021/java-chess/pull/72#discussion_r576231866
반영

Dae-Hwa committed:
Add getRepresentationBy in Color

색깔에 따라 대소문자 구분하여 반환할 수 있도록 추가
https://github.com/codesquad-members-2021/java-chess/pull/72#discussion_r576234029

Dae-Hwa committed:
Refactor getRepresentation in Pawn

Color의 getRepresentation 사용하도록 수정
https://github.com/codesquad-members-2021/java-chess/pull/72#discussion_r576234029

Dae-Hwa committed:
Remove unnecessary default contstructor

https://github.com/codesquad-members-2021/java-chess/pull/72#discussion_r576249705

Dae-Hwa committed:
Create BoardDTO

출력과 이어질 boardDTO 생성
테스트 용이성을 위해 File의 fillWith 메소드 체이닝 추가
https://github.com/codesquad-members-2021/java-chess/pull/72#discussion_r576252174

Dae-Hwa committed:
Add toDTO in Board

https://github.com/codesquad-members-2021/java-chess/pull/72#discussion_r576252174

Dae-Hwa committed:
Add tests use toDTO in Board to BoardDTOTest

initialze와 toDTO를 사용하여 출력 테스트
테스트 용이성을 위해 Board의 initialze에 메소드 체이닝 추가
https://github.com/codesquad-members-2021/java-chess/pull/72#discussion_r576252174

Dae-Hwa committed:
Refactor Main to use BoardDTO for print

https://github.com/codesquad-members-2021/java-chess/pull/72#discussion_r576252174

Dae-Hwa committed:
Remove print that not needed anymore

Dae-Hwa committed:
Add get to File that use normal index is begin at int 0

char a~h를 사용하는 get 외에 일반적인 인덱스를 사용하는 get 하나 더 추가

Dae-Hwa committed:
Fix I/O Stream not closed to use try-with-resource in Main

Dae-Hwa committed:
Remove unecessary import in Board

Dae-Hwa committed:
Refactor factory method in Pawn for return static color

Dae-Hwa committed:
Remove getPawnsResult that unecessary

ksundong commented on src/main/java/kr/codesquad/freddie/chess/Main.java:
```diff
@@ -0,0 +1,24 @@
+
+public class Main {
+    public static void main(String[] args) throws IOException {
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in))) {
```
`BufferedReader` 부분 유지하셔도 됩니다~
try-with-resources는 잘 사용하셨어요!

ksundong commented on src/main/java/kr/codesquad/freddie/chess/Main.java:
```diff
@@ -0,0 +1,24 @@
+                 command = br.readLine()) {
+                if (command.equals("start")) {
+                    board.initialize();
+                    System.out.println(board.toDTO().getRepresentation());
```
이 부분은 DTO부분은 제거해주시고 원래의 `print()`메서드를 좀 더 개선해줍시다.

ksundong replied:
굳이 DTO 변환할 필요 없이 표현을 board가 가지면 되겠죠?

Dae-Hwa replied:
수정했습니다!

ksundong commented on src/main/java/kr/codesquad/freddie/chess/board/BoardDTO.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class BoardDTO {
```
일단 이 클래스는 제거하고 진행해주세요~

Dae-Hwa replied:
제거했습니다

ksundong commented on src/main/java/kr/codesquad/freddie/chess/piece/Pawn.java:
```diff
@@ -13,14 +13,22 @@ public Pawn(Color color) {
 
-    public static Pawn create(String color) {
-        return new Pawn(Color.valueOf(color.toUpperCase()));
+    public static Pawn createBlackPawn() {
+        return new Pawn(Color.BLACK);
+    }
+
+    public static Pawn createWhitePawn() {
+        return new Pawn(Color.WHITE);
     }
```
👍

ksundong commented on src/main/java/kr/codesquad/freddie/chess/piece/Color.java:
```diff
@@ -1,5 +1,17 @@
 
 public enum Color {
-    WHITE, BLACK;
+    WHITE {
+        @Override
+        String getRepresentationBy(String representation) {
+            return representation.toLowerCase();
+        }
+    }, BLACK {
+        @Override
+        String getRepresentationBy(String representation) {
+            return representation.toUpperCase();
+        }
+    };
```
그냥 representation을 Enum에서 관리하는건 어땠을까요?

Dae-Hwa replied:
Color는 색깔에 관련된 것만 가지고 있는게 더 명확할거라 생각했습니다. 또한, representation은 굳이 관리하자면 Pawn에서 직접 관리하는게 더 좋을 것이라 생각했습니다.

ksundong replied:
좋은 판단이네요. 저도 한 번 다시 생각해보게되었습니다.

ksundong commented on src/test/java/kr/codesquad/freddie/chess/board/BoardDTOTest.java:
```diff
@@ -0,0 +1,110 @@
+package kr.codesquad.freddie.chess.board;
+
+import kr.codesquad.freddie.chess.piece.Color;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.*;
+
+class BoardDTOTest {
```
만드느라 고생하셨겠네요 흑흑...

Dae-Hwa replied:
그래도 테스트를 잘? 짜둔 탓인지 Board의 테스트로 그대로 옮길 수 있었습니다! ㅎㅎ

ksundong reviewed (changes requested):
제가 뜬구름 잡는 얘기를 좀 했나보군요. 죄송합니다.
아마 제 의도가 명확히 전달이 안되었나봐요. 조심해야겠네요.

이미 대부분 깔끔하게 잘 작성해주셔서 DTO 부분 제거만 해주시면 될 것 같아요.
고생하셨습니다~

Dae-Hwa committed:
Remove BoardDTO

BoardDTO 제거, getRepresentation을 Board에 추가

Dae-Hwa commented:
DTO 제거했습니다. print는 진짜 print를 해줘야할 것 같은 느낌이라, 대신 getRepresentation이라 명명하여 Board에 다시 추가해줬습니다.
제가 리뷰를 너무 어렵게 받아들였나 봐요 ㅠㅠ 다음부터 좀 더 차근차근 읽어보겠습니다! 
어찌됐건 저 나름의 생각을 많이 해보게 돼서 좋았던 것 같아요! 특히 BufferedReader같은 경우에는 마냥 성능이 더 좋으니까 혹은 다루기 편해서 정도로 생각하고 사용하고 있었는데 이런 식으로 접근해보니 재밌네요 ㅎㅎ 좋은 리뷰 감사합니다👍

ksundong reviewed (approved):
고생하셨습니다! 다음단계 진행해주세요!
